core/src/main/scala/scalaz/IList.scala

```diff
@@ -697,10 +697,19 @@ private trait IListEqual[A] extends Equal[IList[A]] {
   implicit def A: Equal[A]
 
   @tailrec final override def equal(a: IList[A], b: IList[A]): Boolean =
-    (a, b) match {
-      case (INil(), INil()) => true
-      case (ICons(a, as), ICons(b, bs)) if A.equal(a, b) => equal(as, bs)
-      case _ => false
+    (a.asInstanceOf[AnyRef] eq b.asInstanceOf[AnyRef]) || {
+      a match {
+        case _: INil[A] => b match {
+          case _ : INil[A] => true
+          case _ => false
+        }
+        case ac: ICons[A] => b match {
+          case _ : INil[A] => false
+          case bc: ICons[A] =>
+            (ac.head.asInstanceOf[AnyRef] eq bc.head.asInstanceOf[AnyRef]) ||
```

Comment:
Eventually we should push this into the `A`, i.e. fix all the other `Equal` in the world. But works for now. 👍 

The only other thing I'd do is swap the order of nil and cons in this second pattern match, since the more common case in comparing two arbitrary lists will be cons and you can skip the initial check for nil.